simpeg/directives/_regularization.py

```diff
@@ -0,0 +1,494 @@
+    def cooling_rate(self, value):
+        self._cooling_rate = validate_integer("cooling_rate", value, min_val=1)
+
+    @property
+    def irls_cooling_factor(self) -> float:
+        """IRLS threshold parameter (epsilon) is divided by this value every iteration."""
+        return self._irls_cooling_factor
+
+    @irls_cooling_factor.setter
+    def irls_cooling_factor(self, value):
+        self._irls_cooling_factor = validate_float(
+            "irls_cooling_factor", value, min_val=0.0, inclusive_min=False
+        )
+
+    @property
+    def f_min_change(self) -> float:
+        """Target chi factor to start the IRLS process."""
+        return self._f_min_change
+
+    @f_min_change.setter
+    def f_min_change(self, value):
+        self._f_min_change = validate_float(
+            "f_min_change", value, min_val=0, inclusive_min=False
+        )
+
+    def misfit_from_chi_factor(self, chi_factor: float) -> float:
+        """
+        Compute the target misfit from the chi factor.
+
+        Parameters
+        ----------
+        chi_factor : float
+            Chi factor to compute the target misfit from.
+        """
+        value = 0
+        if isinstance(self.survey, list):
+            for survey in self.survey:
+                value += survey.nD * chi_factor
+
+        else:
+            value = self.survey.nD * chi_factor
+
+        return value
+
+    def adjust_cooling_schedule(self):
+        """
+        Adjust the cooling schedule based on the misfit.
+        """
+        ratio = self.invProb.phi_d / self.misfit_from_chi_factor(self.chifact_target)
+        if np.abs(1.0 - ratio) > self.misfit_tolerance:
+
+            if ratio > 1:
+                ratio = np.mean([2.0, ratio])
+            else:
+                ratio = np.mean([0.75, ratio])
+
+            self.cooling_factor = ratio
+
+    def initialize(self):
+        """
+        Initialize the IRLS iterations with l2-norm regularization (mode:1).
+        """
+
+        input_norms = []
+        for reg in self.reg.objfcts:
+            if not isinstance(reg, Sparse):
+                continue
+
+            input_norms += [reg.norms]
+            reg.norms = [2.0 for _ in reg.objfcts]
+
+        self._metrics = IRLSMetrics(input_norms=input_norms)
+
+    def endIter(self):
+        """
+        Check on progress of the inversion and start/update the IRLS process.
+        """
+        # After reaching target misfit with l2-norm, switch to IRLS (mode:2)
+        if (
+            self.metrics.start_irls_iter is None
+            and self.invProb.phi_d < self.misfit_from_chi_factor(self.chifact_start)
+        ):
+            self.start_irls()
+
+        # Check if misfit is within the tolerance, otherwise scale beta
+        self.adjust_cooling_schedule()
+
+        # Only update after GN iterations
+        if (
+            self.metrics.start_irls_iter is not None
+            and (self.opt.iter - self.metrics.start_irls_iter) % self.cooling_rate == 0
```

Comment:
From what I understood from the docstrings, the `cooling_rate` was used to control how often we update `beta`. But here it's used to also control how often we update the `irls_threshold`s. Is this the desired behaviour? If that's so, could we make it more explicit in the docs?